extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js

```diff
@@ -0,0 +1,219 @@
+ */
+
+/**
+ * @implements {../../../src/video-interface.VideoInterface}
+ */
+
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {user} from '../../../src/log';
+import {
+  installVideoManagerForDoc,
+} from '../../../src/service/video-manager-impl';
+import {removeElement} from '../../../src/dom';
+import {isObject} from '../../../src/types';
+import {tryParseJson} from '../../../src/json';
+import {VideoEvents} from '../../../src/video-interface';
+import {videoManagerForDoc} from '../../../src/video-manager';
+
+class AmpNexxtvPlayer extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Element} */
+    this.iframe_ = null;
+
+    /** @private {?Promise} */
+    this.playerReadyPromise_ = null;
+
+    /** @private {?Function} */
+    this.playerReadyResolver_ = null;
+  }
+
+  /**
+   * @param {boolean=} opt_onLayout
+   * @override
+   */
+  preconnectCallback(opt_onLayout) {
+    this.preconnect.url(this.origin_, opt_onLayout);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  buildCallback() {
+    this.playerReadyPromise_ = new Promise(resolve => {
+      this.playerReadyResolver_ = resolve;
+    });
+
+    this.mediaid_ = user().assert(
+      this.element.getAttribute('data-mediaid'),
+      'The data-mediaid attribute is required for <amp-nexxtv-player> %s',
+      this.element);
+
+    this.client_ = user().assert(this.element.getAttribute('data-client'),
+      'The data-client attribute is required for <amp-nexxtv-player> %s',
+      this.element);
+
+    this.start_ = this.element.getAttribute('data-seek-to') || 0;
+    this.mode_ = this.element.getAttribute('data-mode') || 'static';
+    this.streamtype_ = this.element.getAttribute('data-streamtype') || 'video';
+    this.origin_ = this.element.getAttribute('data-origin')
+                  || 'https://embed.nexx.cloud/';
+
+    const iframe = this.element.ownerDocument.createElement('iframe');
+    this.iframe_ = iframe;
+
+    this.forwardEvents([VideoEvents.PLAY, VideoEvents.PAUSE], iframe);
+    this.applyFillContent(iframe);
+    iframe.setAttribute('frameborder', '0');
+    iframe.setAttribute('allowfullscreen', 'true');
+
+    this.element.appendChild(iframe);
+
+    installVideoManagerForDoc(this.element);
+    videoManagerForDoc(this.element).register(this);
+  }
+
+  /** @override */
+  viewportCallback(visible) {
+    this.element.dispatchCustomEvent(VideoEvents.VISIBILITY, {visible});
+  }
+
+  /** @override */
+  layoutCallback() {
+    let src = this.origin_;
+
+    if (this.streamtype_ !== 'video') {
+      src += `${encodeURIComponent(this.streamtype_)}/`;
+    }
+
+    src += `${encodeURIComponent(this.client_)}/`;
+    src += `${encodeURIComponent(this.mediaid_)}`;
+    src += `?start=${encodeURIComponent(this.start_)}`;
+    src += `&datamode=${encodeURIComponent(this.mode_)}&amp=1`;
```

Comment:
Move this logic to a `getVideoIframeSrc_` method that is called here and in `preconnectCallback`